servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLRequestHandler.kt

```diff
@@ -16,37 +16,136 @@
+import com.expediagroup.graphql.server.types.GraphQLBatchResponse
 import com.expediagroup.graphql.server.types.GraphQLRequest
 import com.expediagroup.graphql.server.types.GraphQLResponse
+import com.expediagroup.graphql.server.types.GraphQLServerRequest
+import com.expediagroup.graphql.server.types.GraphQLServerResponse
+import graphql.ExecutionInput
 import graphql.GraphQL
+import graphql.execution.instrumentation.ChainedInstrumentation
+import graphql.execution.instrumentation.Instrumentation
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
 import kotlinx.coroutines.future.await
+import kotlinx.coroutines.supervisorScope
 
 open class GraphQLRequestHandler(
     private val graphQL: GraphQL,
-    private val dataLoaderRegistryFactory: DataLoaderRegistryFactory? = null
+    private val dataLoaderRegistryFactory: KotlinDataLoaderRegistryFactory? = null
 ) {
 
+    private val batchingInstrumentationStrategy: Class<out Instrumentation>? =
+        (graphQL.instrumentation as? ChainedInstrumentation)?.let { chainedInstrumentation ->
+            chainedInstrumentation
+                .instrumentations
+                .firstOrNull { instrumentation ->
+                    instrumentation.javaClass == DataLoaderLevelDispatchedInstrumentation::class.java ||
+                        instrumentation.javaClass == DataLoaderSyncExecutionExhaustedInstrumentation::class.java
+                }
+                ?.javaClass
+        }
+
     /**
      * Execute a GraphQL request in a non-blocking fashion.
      * This should only be used for queries and mutations.
      * Subscriptions require more specific server logic and will need to be handled separately.
      */
-    open suspend fun executeRequest(request: GraphQLRequest, context: GraphQLContext? = null, graphQLContext: Map<*, Any> = emptyMap<Any, Any>()): GraphQLResponse<*> {
-        // We should generate a new registry for every request
-        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
-        val executionInput = request.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
+    open suspend fun executeRequest(
+        graphQLRequest: GraphQLServerRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>()
+    ): GraphQLServerResponse =
+        when (graphQLRequest) {
+            is GraphQLRequest -> {
+                GraphQLResponse(
+                    execute(
+                        graphQLRequest.toExecutionInput(context, dataLoaderRegistryFactory?.generate(), graphQLContext)
+                    )
+                )
+            }
+            is GraphQLBatchRequest -> {
+                when {
+                    graphQLRequest.requests.none(GraphQLRequest::isMutation) -> {
+                        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
+                        val contextWithDataLoaderInstrumentationState = graphQLContext + (
```

Comment:
this part is tricky, given that at this layer we don't have access to the configuration, we need to check the `GraphQL` instance instrumentations to check if there is a `DataLoaderLevelDispatchedInstrumentation` or `DataLoaderSyncExecutionExhaustedInstrumentation`, if so, it means that batching is enabled and we need to add the corresponding `State` instance to the `GraphQLContext` before attempting to execute the operations in a concurrent way.

The DataLoaderRegistry needs to be created only once for all operations in a `GraphQLBatchRequest`